[PLAT-435] Dataverse Refactor

Johnetordoff opened the pull request:
## Purpose

Allows us to access Dataverse datasets via v2 API.  Also there were some incidental style fixes which a threw in rather then make a new ticket.

## Changes


Refactor:
- add get_folder method to model
- adapt view to use model method
- refactor front end to use "name" instead of "title", in accordance with API standard

Style:
- Fix broken Dataverse and dataset link
- Change text color from blue to black

## QA Notes

Style changes below.

<img width="909" alt="screen shot 2018-03-26 at 12 07 23 pm" src="https://user-images.githubusercontent.com/9688518/37918361-d25d27d6-30ee-11e8-9f9d-49ba7358a2cb.png">

## Side Effects

Because the the Dataverse Repo -> Dataverse -> Dataset -> File sturcture of Dataverse, this needs a degree more configuration from the other addons. So to get Datasets to appear via the API you have to select a Dataverse with a Dataset via the UI.

## Ticket

https://openscience.atlassian.net/projects/PLAT/issues/PLAT-435

pattisdr commented on addons/dataverse/models.py:
```diff
@@ -125,6 +125,24 @@ def nodelogger(self):
+        connection = client.connect_from_settings(self)
+        dataverse = client.get_dataverse(connection, self.dataverse_alias)
+
+        datasets = client.get_datasets(dataverse)
```
Regarding my work on refactoring citations for API v2 (https://github.com/CenterForOpenScience/osf.io/pull/8230) Zotero also has two (and sometimes more) levels,  You first pick your library, which can be a group library or your personal library, which I think can be compared to your dataverse.  Then from that library, you select your collection, which can be equated to a dataset. 

How I structured the API folders endpoint for Zotero, was top-level was the library.  And children of that library were collections, and then children of those collections were more collections (since you can have collections within collections).   My API can go multiple levels.  

Is this what is wanted for Dataverse? Top level get_folders returns dataverses, and then the children link returns datasets.  (Granted, my PR is in Final Review right now, and I may have misunderstood what was wanted here, so check with @sloria.) 

Edit: Perhaps Zotero is different, in that library selection and folder selection are two explicit steps, saved separately, and dataverse it looks like a 1-step thing.  For Zotero, you can have a library saved, but no collection.

Johnetordoff replied:
I talked to @sloria about this and it's going to require multiple steps. I think we can still get this step out though.

pattisdr commented on addons/dataverse/models.py:
```diff
@@ -125,6 +125,24 @@ def nodelogger(self):
+        connection = client.connect_from_settings(self)
+        dataverse = client.get_dataverse(connection, self.dataverse_alias)
+
+        datasets = client.get_datasets(dataverse)
```
Do we need error handling if client.get_datasets(dataverse) fails or client.get_dataverse(...) fails somehow?

pattisdr commented on addons/dataverse/static/dataverseNodeConfig.js:
```diff
@@ -253,7 +253,7 @@ ViewModel.prototype.updateFromData = function(data) {
         self.selectedDataverseAlias(data.savedDataverse.alias);
         self.savedDatasetDoi(data.savedDataset.doi);
-        self.savedDatasetTitle(data.savedDataset.title);
+        self.savedDatasetTitle(data.savedDataset.name);
```
If you connect a dataverse, and then connect a dataset and save, then refresh the page, I don't know if savedDatasetTitle is null or what, but the dataset title is not showing up.  If you connect one it shows, but not on first load.

![screen shot 2018-04-25 at 10 26 14 am](https://user-images.githubusercontent.com/9755598/39255919-3ec3cccc-4873-11e8-864f-80ab0b96b873.png)

pattisdr commented on addons/dataverse/static/dataverseNodeConfig.js:
```diff
@@ -132,10 +132,10 @@ function ViewModel(url) {
 
     self.savedDatasetUrl = ko.pureComputed(function() {
-        return (self.urls()) ? self.urls().datasetPrefix + self.savedDatasetDoi() : null;
+        return (self.urls()) ? self.urls().dataversePrefix + 'dataset.xhtml?persistentId=' + self.savedDatasetDoi() : null;
```
Nice catch on these URL's 👍

pattisdr commented on addons/dataverse/views.py:
```diff
@@ -150,16 +150,8 @@ def dataverse_set_config(node_addon, auth, **kwargs):
-    }
-    return ret, http.OK
+    node_addon.dataverse_alias = request.json.get('alias')
+    return {'alias': node_addon.dataverse_alias, 'datasets': node_addon.get_folders()}, http.OK
```
Minor: probably don't need to define node_addon.dataverse_alias here, can just call request.json.get('alias') where you are returning this dictionary? Or is there a reason why you are saving a dataverse alias under node_addon?

pattisdr commented on addons/dataverse/templates/dataverse_node_settings.mako:
```diff
@@ -88,12 +88,12 @@
                                 </div>
                                 <div data-bind="if: showNoDatasets">
-                                    <div class="text-info" style="padding-top: 8px">
+                                    <div class="text-muted" style="padding-top: 8px">
```
👍 nice, I had clicked on this in the past thinking it was a link.

Johnetordoff commented:
@pattisdr I'm sorry, it's probably very confusing to track all the changes, but I've changed the code to list dataverses in get_folder instead of datasets because I feel that's more in accordance with the use case. I'm keeping the old get_folder method is renamed get_dataset and is still being used the same way in the view, we'll likely v2-ize it eventually once we work out the dataverse v2 process a bit more.

 I also fixed that bug you mentioned.

pattisdr commented on addons/dataverse/models.py:
```diff
@@ -125,6 +125,46 @@ def nodelogger(self):
+        connection = client.connect_from_settings(self)
+
+        dataverses = connection.get_dataverses()
+        print(dataverses)
```
extra print statement

sloria commented on api_tests/nodes/views/test_node_addons.py:
```diff
@@ -685,6 +685,17 @@ def _settings_kwargs(self, node, user_settings):
         }
 
 
+class MockConnection():
```
Minor: since we're on Python 2, "top-level" classes should still inherit `object`. Shouldn't make a difference in this case.

sloria commented:
This is headed to staging3 for QA.
